docs/architecture/widget_framework.md

```diff
@@ -0,0 +1,90 @@
+1. We get the config stored under `id` in the screen configuration file.
+1. We look up the "app ID" stored in the config. We get the appropriate candidate generator module for that app ID.
+1. We get the template for this screen type by calling `candidate_generator.screen_template()`.
+1. We get a list of candidate widgets for the screen by calling `candidate_generator.candidate_instances(config)`.
+1. Candidates are sorted by priority (descending), and then placed into template slots one at a time until the template is filled. We call this populated template a "layout".
+1. Paged sections of the template are pruned to just the current page, based on the current clock time.
+1. Widgets are serialized to JSON-friendly maps and dropped into a map that reflects the chosen layout.
+1. The map is converted to JSON and sent back to the client.
+1. Client receives JSON response and passes it into the React component tree to render updates.
+
+## Data flow for an audio readout request
+
+Audio equivalence support is built into the framework.
+
+Its data flow is about the same as the visual UI, but instead of being serialized to JSON and sent to the browser to be rendered by React, it is instead rendered to an SSML string on the server and piped through AWS Polly to produce an MP3 containing synthesized speech.
+
+1. Client requests `/v2/audio/{id}/readout.mp3`. Subsequent steps take place on the server.
+1. Steps **2** through **8** of the above data flow run as usual.
+1. The data map is rendered to SSML using each widget's audio view.
+1. The SSML is sent to Polly for synthesis, and we receive MP3 data in the response.
+1. The MP3 is sent back to the client for playback.
+
+## Anatomy of a template
+
+Each screen type has a template that defines the "slots" that widgets can fit into, as well as layout variations for special cases like high-priority alerts that take over all space below the screen header.
+
+A template is a recursive data structure.
+
+The base case is a "slot ID"—a simple named space that a widget can drop into.
+
+The recursive case is a tuple representing a region containing one or more layout variations: `{region_id, %{layout_id => list(template)}}`. Let's break that down:
+
+- `region_id`: the region's name. For example: `:body`, `:flex_zone`, `:screen`.
+- `layout_id`: a name for one possible set of slots that can fill this region. For example: `:body_normal`, `:body_takeover`; `:one_large`, `:two_medium`.
+- `list(template)`: the set of slots, or even further nested regions, that fill this region under the given layout variation.
+
+Layout variations provide a mechanism for high-priority widgets, like urgent alerts, to take over large portions of a screen. They also allow paged regions to hold widgets of different sizes per page.
+
+## Feature: Paging
+
+Some screen types have paged regions, where the widgets that appear in the region can rotate over the course of several data refreshes.
```

Comment:
A special case here is the `elevator_status`. Could be confusing to someone new to the framework due to that widget appearing to page the same way the flex zone does.

Reply:
Added a section about persistent widgets / client-side widget-internal paging!